src/synth/fluid_synth_polymono.c

```diff
@@ -0,0 +1,536 @@
+{
+    int i,n_chan;
+    int result;
+    /* checks parameters first */
+    fluid_return_val_if_fail (synth != NULL, FLUID_FAILED);
+    fluid_return_val_if_fail (n >= 0, FLUID_FAILED);
+    fluid_synth_api_enter(synth);
+    
+    n_chan = synth->midi_channels; /* MIDI Channels number */
+    if (n > n_chan)
+    {
+        FLUID_API_RETURN(FLUID_FAILED);
+    }
+    
+    /* Checks if information are valid  */
+    if(n && basicChannelInfos ) for (i = 0; i < n; i++)
+    {
+	if (	basicChannelInfos[i].basicchan < 0 || 
+		basicChannelInfos[i].basicchan >= n_chan ||
+		basicChannelInfos[i].mode < 0 ||
+		basicChannelInfos[i].mode >= FLUID_CHANNEL_MODE_LAST ||
+		basicChannelInfos[i].val < 0 ||
+		basicChannelInfos[i].basicchan + basicChannelInfos[i].val > n_chan)
+        {
+            FLUID_API_RETURN(FLUID_FAILED);
+        }
+    }
+
+    /* Clears previous list of basic channel */
+    for(i = 0; i <  n_chan; i++) {
+	reset_fluid_channel_basic_channel_infos(synth->channel[i]);
+	synth->channel[i]->mode_val = 0; 
+    }
+    if(n && basicChannelInfos)
+    {
+	result = FLUID_OK;
+
+	/* Sets the new list of basic channel */
+	for (i = 0; i < n; i++)
+	{
+	    int basic_chan = basicChannelInfos[i].basicchan;
+        if (synth->channel[basic_chan]->mode &  FLUID_CHANNEL_BASIC)
+		/* Different entries have the same basic channel. 
+		 An entry supersedes a previous entry with the same 
+		 basic channel.*/
+	        FLUID_LOG(FLUID_INFO, warning_msg);
+	    /* Set Basic channel first */
+        else synth->channel[basic_chan]->mode |= FLUID_CHANNEL_BASIC;
+	}
+
+	for (i = 0; i < n; i++)
+	{
+	    int r =fluid_synth_set_basic_channel_LOCAL( synth, 
+				basicChannelInfos[i].basicchan,
+				basicChannelInfos[i].mode,
+				basicChannelInfos[i].val);
+	    if (result == FLUID_OK) result = r;
+	}
+    }
+    else result = fluid_synth_set_basic_channel_LOCAL( synth, 0, FLUID_CHANNEL_MODE_OMNION_POLY,0);
+    FLUID_API_RETURN(result);
+}
+
+/**
+ * Changes the mode of an existing basic channel or inserts a new basic channel part.
+ *
+ * - If basicchan is already a basic channel, the mode is changed.
+ * - If basicchan is not a basic channel, a new basic channel part is inserted between
+ * the previous basic channel and the next basic channel. val value of the previous
+ * basic channel will be narrowed if necessary. 
+ * 
+ * @param synth the synth instance.
+ * @param basicchan the Basic Channel number (0 to MIDI channel count-1).
+ * @param mode the MIDI mode to use for basicchan (0 to 3).
+ * @param val Number of monophonic channels (for mode 3 only) (0 to MIDI channel count).
+ * 
+ * @return 
+ * - FLUID_OK on success.
+ * - FLUID_FAILED
+ *   - synth is NULL.
+ *   - chan or val is outside MIDI channel count.
+ *   - mode is invalid.
+ *   - val has a number of channels overlapping the next basic channel.
+ */
+int fluid_synth_set_basic_channel(fluid_synth_t* synth, int basicchan, int mode, int val)
```

Comment:
Why doesnt `fluid_synth_set_basic_channel()` receive a `fluid_basic_channel_infos_t*`?

Reply:
It could done, if we prefer that the caller build a fluid_basic_channel_infos_t ?

Reply:
I'm not a fan of introducing these helper structs. However in this case I currently cannot think of a better way. And since we already have this `fluid_basic_channel_infos_t`, we should also consistently use it.